lib/awaitableSender.ts

```diff
@@ -187,12 +192,48 @@ export class AwaitableSender extends BaseSender {
           return reject(new OperationTimeoutError(message));
         }, this.sendTimeoutInSeconds * 1000);
 
+        const onAbort = () => {
+          const promise = this.deliveryDispositionMap.get(delivery.id) as PromiseLike;
+          clearTimeout(promise.timer);
+          const deleteResult = this.deliveryDispositionMap.delete(delivery.id);
+          log.sender(
+            "[%s] Event: 'abort', Successfully deleted the delivery with id %d from the " +
+            " map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+            this.connection.id, delivery.id, this.name, this.session.id, deleteResult
+          );
+          const err = new Error("Send request has been cancelled.");
+          err.name = abortErrorName;
+          log.error("[%s] %s", this.connection.id, err.message);
+          promise.reject(err);
+        };
+
+        const removeAbortListener = () => {
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAbort);
+          }
+        };
+
         const delivery = (this._link as RheaSender).send(msg, tag, format);
         this.deliveryDispositionMap.set(delivery.id, {
-          resolve: resolve,
-          reject: reject,
-          timer: timer
+          resolve: () => {
+            resolve();
+            removeAbortListener();
+          },
+          reject: (reason?: any) => {
+            reject(reason);
+            removeAbortListener();
+          },
+          timer: timer,
+          onAbort
         });
+
+        if (abortSignal) {
+          if (abortSignal.aborted) {
+            onAbort();
```

Comment:
It made for clear coding to handle both aborted signal and signal that is fired later using the same callback. I believe this works well in other areas except this one. The difference is that there is different error code path which can be avoided with the abort signal was fired.

But, my follow up would be that if the user is meaning to cancel an operation, does it really matter to them if they hit another code path and got a different error? The idea is to exit as early as possible right?

Reply:
> But, my follow up would be that if the user is meaning to cancel an operation, does it really matter to them if they hit another code path and got a different error? The idea is to exit as early as possible right?

We treat `InsufficientCreditError` as a retryable error in our `@azure/core-amqp` package:
https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-amqp/src/errors.ts#L552-L554

So if there are any retries remaining, we'd attempt another send rather than cancelling the operation.

I also suspect, since we do this in our own code, that users might handle `AbortError` in their code differently than other errors which we'd miss out on in this scenario.

Reply:
Fair point. Fixed in https://github.com/amqp/rhea-promise/pull/66/commits/a18c3b71eb4f25863908f4ecfde973570db0137e